File: sandbox/protocol/binary.ts
/**
 * Binary Reader/Writer for GameProtocol
 *
 * Matches the game's CustomBinaryReader/CustomBinaryWriter format:
 * - Little-endian encoding
 * - Length-prefixed strings (uint16 + utf8 bytes)
 * - Array serialization (uint16 count + items)
 */

// =============================================================================
// BINARY READER
// =============================================================================

export class BinaryReader {
  private view: DataView;
  private offset: number = 0;

  constructor(buffer: ArrayBuffer | Uint8Array) {
    if (buffer instanceof Uint8Array) {
      this.view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    } else {
      this.view = new DataView(buffer);
    }
  }

  get position(): number {
    return this.offset;
  }

  get remaining(): number {
    return this.view.byteLength - this.offset;
  }

  readByte(): number {
    const value = this.view.getUint8(this.offset);
    this.offset += 1;
    return value;
  }

  readBool(): boolean {
    return this.readByte() !== 0;
  }

  readInt16(): number {
    const value = this.view.getInt16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readUInt16(): number {
    const value = this.view.getUint16(this.offset, true);
    this.offset += 2;
    return value;
  }

  readInt32(): number {
    const value = this.view.getInt32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readUInt32(): number {
    const value = this.view.getUint32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readInt64(): bigint {
    const value = this.view.getBigInt64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readUInt64(): bigint {
    const value = this.view.getBigUint64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readFloat(): number {
    const value = this.view.getFloat32(this.offset, true);
    this.offset += 4;
    return value;
  }

  readDouble(): number {
    const value = this.view.getFloat64(this.offset, true);
    this.offset += 8;
    return value;
  }

  readString(): string {
    const length = this.readUInt16();
    if (length === 0) return "";

    const bytes = new Uint8Array(this.view.buffer, this.view.byteOffset + this.offset, length);
    this.offset += length;
    return new TextDecoder("utf-8").decode(bytes);
  }

  readBytes(count: number): Uint8Array {
    const bytes = new Uint8Array(this.view.buffer, this.view.byteOffset + this.offset, count);
    this.offset += count;
    return bytes;
  }

  /** Read array with length prefix and reader function */
  readArray<T>(reader: () => T): T[] {
    const count = this.readUInt16();
    const items: T[] = [];
    for (let i = 0; i < count; i++) {
      items.push(reader());
    }
    return items;
  }
}

// =============================================================================
// BINARY WRITER
// =============================================================================

export class BinaryWriter {
  private buffer: ArrayBuffer;
  private view: DataView;
  private offset: number = 0;
  private capacity: number;

  constructor(initialCapacity: number = 1024) {
    this.capacity = initialCapacity;
    this.buffer = new ArrayBuffer(initialCapacity);
    this.view = new DataView(this.buffer);
  }

  private ensureCapacity(additionalBytes: number): void {
    if (this.offset + additionalBytes > this.capacity) {
      // Double capacity or add enough for the new data
      const newCapacity = Math.max(this.capacity * 2, this.offset + additionalBytes);
      const newBuffer = new ArrayBuffer(newCapacity);
      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer, 0, this.offset));
      this.buffer = newBuffer;
      this.view = new DataView(this.buffer);
      this.capacity = newCapacity;
    }
  }

  get position(): number {
    return this.offset;
  }

  writeByte(value: number): void {
    this.ensureCapacity(1);
    this.view.setUint8(this.offset, value & 0xff);
    this.offset += 1;
  }

  writeBool(value: boolean): void {
    this.writeByte(value ? 1 : 0);
  }

  writeInt16(value: number): void {
    this.ensureCapacity(2);
    this.view.setInt16(this.offset, value, true);
    this.offset += 2;
  }

  writeUInt16(value: number): void {
    this.ensureCapacity(2);
    this.view.setUint16(this.offset, value, true);
    this.offset += 2;
  }

  writeInt32(value: number): void {
    this.ensureCapacity(4);
    this.view.setInt32(this.offset, value, true);
    this.offset += 4;
  }

  writeUInt32(value: number): void {
    this.ensureCapacity(4);
    this.view.setUint32(this.offset, value, true);
    this.offset += 4;
  }

  writeInt64(value: bigint): void {
    this.ensureCapacity(8);
    this.view.setBigInt64(this.offset, value, true);
    this.offset += 8;
  }

  writeUInt64(value: bigint): void {
    this.ensureCapacity(8);
    this.view.setBigUint64(this.offset, value, true);
    this.offset += 8;
  }

  writeFloat(value: number): void {
    this.ensureCapacity(4);
    this.view.setFloat32(this.offset, value, true);
    this.offset += 4;
  }

  writeDouble(value: number): void {
    this.ensureCapacity(8);
    this.view.setFloat64(this.offset, value, true);
    this.offset += 8;
  }

  writeString(value: string | null): void {
    if (value === null || value === undefined) {
      this.writeUInt16(0);
      return;
    }
    const bytes = new TextEncoder().encode(value);
    this.writeUInt16(bytes.length);
    this.writeBytes(bytes);
  }

  writeBytes(bytes: Uint8Array): void {
    this.ensureCapacity(bytes.length);
    new Uint8Array(this.buffer, this.offset, bytes.length).set(bytes);
    this.offset += bytes.length;
  }

  /** Write array with length prefix and writer function */
  writeArray<T>(items: T[] | null, writer: (item: T) => void): void {
    if (items === null || items === undefined) {
      this.writeUInt16(0);
      return;
    }
    this.writeUInt16(items.length);
    for (const item of items) {
      writer(item);
    }
  }

  /** Get the written bytes as a new Uint8Array */
  toBytes(): Uint8Array {
    return new Uint8Array(this.buffer, 0, this.offset);
  }

  /** Get the written bytes as an ArrayBuffer */
  toArrayBuffer(): ArrayBuffer {
    return this.buffer.slice(0, this.offset);
  }
}

File: server-py/protocol/binary.py
"""
Binary Reader/Writer for GameProtocol

Matches the game's CustomBinaryReader/CustomBinaryWriter format:
- Little-endian encoding
- Length-prefixed strings (uint16 + utf8 bytes)
- Array serialization (uint16 count + items)
"""

import struct
from typing import Callable, TypeVar, List, Optional
from io import BytesIO

T = TypeVar("T")


class BinaryReader:
    """Binary reader with little-endian encoding"""

    def __init__(self, data: bytes):
        self._buffer = BytesIO(data)
        self._length = len(data)

    @property
    def position(self) -> int:
        return self._buffer.tell()

    @property
    def remaining(self) -> int:
        return self._length - self._buffer.tell()

    def read_byte(self) -> int:
        """Read unsigned 8-bit integer"""
        return struct.unpack("<B", self._buffer.read(1))[0]

    def read_bool(self) -> bool:
        """Read boolean (1 byte)"""
        return self.read_byte() != 0

    def read_int16(self) -> int:
        """Read signed 16-bit integer"""
        return struct.unpack("<h", self._buffer.read(2))[0]

    def read_uint16(self) -> int:
        """Read unsigned 16-bit integer"""
        return struct.unpack("<H", self._buffer.read(2))[0]

    def read_int32(self) -> int:
        """Read signed 32-bit integer"""
        return struct.unpack("<i", self._buffer.read(4))[0]

    def read_uint32(self) -> int:
        """Read unsigned 32-bit integer"""
        return struct.unpack("<I", self._buffer.read(4))[0]

    def read_int64(self) -> int:
        """Read signed 64-bit integer"""
        return struct.unpack("<q", self._buffer.read(8))[0]

    def read_uint64(self) -> int:
        """Read unsigned 64-bit integer"""
        return struct.unpack("<Q", self._buffer.read(8))[0]

    def read_float(self) -> float:
        """Read 32-bit float"""
        return struct.unpack("<f", self._buffer.read(4))[0]

    def read_double(self) -> float:
        """Read 64-bit float"""
        return struct.unpack("<d", self._buffer.read(8))[0]

    def read_string(self) -> str:
        """Read length-prefixed UTF-8 string"""
        length = self.read_uint16()
        if length == 0:
            return ""
        data = self._buffer.read(length)
        return data.decode("utf-8")

    def read_bytes(self, count: int) -> bytes:
        """Read raw bytes"""
        return self._buffer.read(count)

    def read_array(self, reader: Callable[[], T]) -> List[T]:
        """Read array with length prefix and reader function"""
        count = self.read_uint16()
        return [reader() for _ in range(count)]


class BinaryWriter:
    """Binary writer with little-endian encoding"""

    def __init__(self, initial_capacity: int = 1024):
        self._buffer = BytesIO()

    @property
    def position(self) -> int:
        return self._buffer.tell()

    def write_byte(self, value: int) -> None:
        """Write unsigned 8-bit integer"""
        self._buffer.write(struct.pack("<B", value & 0xFF))

    def write_bool(self, value: bool) -> None:
        """Write boolean (1 byte)"""
        self.write_byte(1 if value else 0)

    def write_int16(self, value: int) -> None:
        """Write signed 16-bit integer"""
        self._buffer.write(struct.pack("<h", value))

    def write_uint16(self, value: int) -> None:
        """Write unsigned 16-bit integer"""
        self._buffer.write(struct.pack("<H", value))

    def write_int32(self, value: int) -> None:
        """Write signed 32-bit integer"""
        self._buffer.write(struct.pack("<i", value))

    def write_uint32(self, value: int) -> None:
        """Write unsigned 32-bit integer"""
        self._buffer.write(struct.pack("<I", value))

    def write_int64(self, value: int) -> None:
        """Write signed 64-bit integer"""
        self._buffer.write(struct.pack("<q", value))

    def write_uint64(self, value: int) -> None:
        """Write unsigned 64-bit integer"""
        self._buffer.write(struct.pack("<Q", value))

    def write_float(self, value: float) -> None:
        """Write 32-bit float"""
        self._buffer.write(struct.pack("<f", value))

    def write_double(self, value: float) -> None:
        """Write 64-bit float"""
        self._buffer.write(struct.pack("<d", value))

    def write_string(self, value: Optional[str]) -> None:
        """Write length-prefixed UTF-8 string"""
        if value is None:
            self.write_uint16(0)
            return
        data = value.encode("utf-8")
        self.write_uint16(len(data))
        self._buffer.write(data)

    def write_bytes(self, data: bytes) -> None:
        """Write raw bytes"""
        self._buffer.write(data)

    def write_array(
        self, items: Optional[List[T]], writer: Callable[[T], None]
    ) -> None:
        """Write array with length prefix and writer function"""
        if items is None:
            self.write_uint16(0)
            return
        self.write_uint16(len(items))
        for item in items:
            writer(item)

    def to_bytes(self) -> bytes:
        """Get the written bytes"""
        return self._buffer.getvalue()
